File: typescript/test/approval.test.ts
/**
 * Tests for the approval_request content part + approve/deny control kinds.
 *
 * Mirrors go/approval_test.go and python/tests/test_approval.py, and exercises
 * the resolve path (part_appended pending → part_updated status flip).
 */
import {describe, expect, it} from 'vitest';
import {
    fromMemoryLayerMessage,
    reduceEvents,
    toMemoryLayerPayload,
    type ApprovalRequestPart,
    type ChatMessage,
    type ControlPart,
    type MessageStartedEvent,
    type PartAppendedEvent,
    type PartUpdatedEvent,
} from '../src/index';

function approvalPart(): ApprovalRequestPart {
    return {
        type: 'approval_request',
        id: 'appr_1',
        tool: 'shell',
        summary: 'Run: npm test',
        args: {command: 'npm test'},
        options: ['once', 'session', 'always'],
        status: 'pending',
        reason: 'shell is not pre-authorized',
    };
}

describe('approval_request part', () => {
    it('resolves live: part_appended (pending) then part_updated flips status', () => {
        const msg: ChatMessage = {
            schema_version: '1.0', id: 'm1', role: 'assistant',
            addr: {thread_id: 'thr'}, meta: {}, content: [],
        };
        const state = reduceEvents([
            {event: 'message_started', message: msg} as MessageStartedEvent,
            {event: 'part_appended', message_id: 'm1', index: 0, part: approvalPart()} as PartAppendedEvent,
            {event: 'part_updated', message_id: 'm1', index: 0, patch: {status: 'approved'}} as PartUpdatedEvent,
        ]);
        const part = state['m1']!.content[0] as ApprovalRequestPart;
        expect(part.type).toBe('approval_request');
        expect(part.status).toBe('approved');
        expect(part.tool).toBe('shell'); // other fields preserved by shallow merge
    });

    it('round-trips through the MemoryLayer codec', () => {
        const msg: ChatMessage = {
            schema_version: '1.0', id: 'm1', role: 'assistant',
            addr: {thread_id: 'thr'}, meta: {}, content: [approvalPart()],
        };
        const payload = toMemoryLayerPayload(msg);
        const rec = JSON.parse(JSON.stringify({
            id: 'ml', thread_id: 'thr', role: payload.role,
            content: payload.content, metadata: payload.metadata,
        }));
        const back = fromMemoryLayerMessage(rec);
        const part = back.content[0] as ApprovalRequestPart;
        expect(part.type).toBe('approval_request');
        expect(part.id).toBe('appr_1');
        expect(part.status).toBe('pending');
    });

    it('control approve carries request_id + scope', () => {
        const ctrl: ControlPart = {
            type: 'control', kind: 'approve', request_id: 'appr_1', scope: 'session',
        };
        expect(ctrl.kind).toBe('approve');
        expect(ctrl.request_id).toBe('appr_1');
        expect(ctrl.scope).toBe('session');
    });
});

File: go/approval_test.go
package spec

import (
	"encoding/json"
	"testing"
)

func Test_ApprovalRequestPart_construct_decode_and_codec(t *testing.T) {
	part := NewApprovalRequestPart(ApprovalRequestPart{
		ID:      "appr_1",
		Tool:    "shell",
		Summary: "Run: npm test",
		Args:    json.RawMessage(`{"command":"npm test"}`),
		Options: []string{"once", "session", "always"},
		Reason:  "shell is not pre-authorized",
	})
	if part.Type() != PartApprovalRequest {
		t.Fatalf("type = %q", part.Type())
	}
	body, ok := part.AsApprovalRequest()
	if !ok {
		t.Fatalf("AsApprovalRequest failed: %s", part.Raw())
	}
	if body.Tool != "shell" || body.Status != ApprovalPending {
		t.Fatalf("body = %#v", body)
	}
	if len(body.Options) != 3 {
		t.Fatalf("options = %#v", body.Options)
	}
	var args map[string]string
	if err := json.Unmarshal(body.Args, &args); err != nil || args["command"] != "npm test" {
		t.Fatalf("args = %s (err %v)", body.Args, err)
	}

	// Survives the MemoryLayer codec (non-text part → data) round-trip.
	msg := ChatMessage{SchemaVersion: "1.0", ID: "m1", Role: RoleAssistant, Addr: MessageAddress{ThreadID: "thr"}, Content: []ContentPart{part}}
	payload, err := ToMemoryLayerPayload(msg)
	if err != nil {
		t.Fatalf("to payload: %v", err)
	}
	raw, _ := json.Marshal(map[string]any{"id": "ml", "thread_id": "thr", "role": payload["role"], "content": payload["content"], "metadata": payload["metadata"]})
	var generic map[string]any
	if err := json.Unmarshal(raw, &generic); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	back, err := FromMemoryLayerMessage(generic)
	if err != nil {
		t.Fatalf("from message: %v", err)
	}
	rt, ok := back.Content[0].AsApprovalRequest()
	if !ok || rt.ID != "appr_1" || rt.Tool != "shell" || rt.Status != ApprovalPending {
		t.Fatalf("approval part lost through codec: %#v", rt)
	}
}

func Test_ControlPart_carries_approve_with_request_id_and_scope(t *testing.T) {
	// The inbound approve/deny control rides request_id + scope.
	raw := []byte(`{"type":"control","kind":"approve","request_id":"appr_1","scope":"session"}`)
	part, err := RawPart(raw)
	if err != nil {
		t.Fatalf("raw part: %v", err)
	}
	body, ok := part.AsControl()
	if !ok {
		t.Fatalf("AsControl failed")
	}
	if body.Kind != ControlApprove || body.RequestID != "appr_1" || body.Scope != "session" {
		t.Fatalf("control body = %#v", body)
	}
}
